x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/execution/search/extractor/AbstractFieldHitExtractor.java

```diff
@@ -147,36 +152,49 @@ private Object extractNestedField(SearchHit hit) {
         return value;
     }
 
-    protected Object unwrapFieldsMultiValue(Object values) {
+    protected Object unwrapFieldsMultiValue(Object values, boolean allowMultivalue) {
         if (values == null) {
             return null;
         }
         if (values instanceof Map && hitName != null) {
             // extract the sub-field from a nested field (dep.dep_name -> dep_name)
-            return unwrapFieldsMultiValue(((Map<?, ?>) values).get(fieldName.substring(hitName.length() + 1)));
+            return unwrapFieldsMultiValue(((Map<?, ?>) values).get(fieldName.substring(hitName.length() + 1)), allowMultivalue);
         }
         if (values instanceof List<?> list) {
             if (list.isEmpty()) {
                 return null;
             } else {
                 if (isPrimitive(list) == false) {
                     if (list.size() == 1 || arrayLeniency) {
-                        return unwrapFieldsMultiValue(list.get(0));
+                        return unwrapFieldsMultiValue(list.get(0), allowMultivalue);
+                    } else if (allowMultivalue) {
+                        values = ((List<?>) values).stream()
+                            .map(x -> unwrapFieldsMultiValue(x, false)) // no arrays of arrays
+                            .collect(Collectors.toList());
                     } else {
                         throw new QlIllegalArgumentException("Arrays (returned by [{}]) are not supported", fieldName);
                     }
                 }
             }
         }
 
-        Object unwrapped = unwrapCustomValue(values);
-        if (unwrapped != null) {
+        Object unwrapped = null;
+        if (values instanceof List<?>) {
+            unwrapped = ((List<?>) values).stream().map(x -> unwrapCustomValue(x)).filter(Objects::nonNull).collect(Collectors.toList());
```

Comment:
1. Not a fan of streams inside tight loops and over small lists.
2. unwrapCustomValue returns null; isn't that a valid value?